### DISCUSSION TITLE
Message data must be a one-dimensional array of string(able) keys and values

### DISCUSSION THREAD

**Discussion by nomaam** [kreait/firebase-php#533] [Q&A]
$data = [
            'title' => $title,
            'body' => $body,
            "ledColor" => [0, 0, 255, 0],
            'actions' => [
                'title' => "test",
                'callback' => 'test',
                'foreground' => "true"
            ]
        ];

    $message = CloudMessage::withTarget('token', $deviceToken)
            ->withData($data);

error

    Message data must be a one-dimensional array of string(able) keys and values.

How do I include an array of values for "actions" and "ledcolor" as it will only accept a one dimension array?

**Comment by jeromegamez** [comment]
Just to be sure that we are talking about the same: this limitation is in the SDK to prevent errors or unexpected results _after_ sending a message to the Firebase FCM API (see https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#Message.FIELDS.data, where it's stated that `data` is a `map (key: string, value: string)`

What you could do is to `json_encode()` the nested arrays before sending the message and then decode it again on the client side:

```php
$data = [
    'title' => $title,
    'body' => $body,
    "ledColor" => json_encode([0, 0, 255, 0]),
    'actions' => json_encode([
        'title' => "test",
        'callback' => 'test',
        'foreground' => "true"
    ])
];
```

**Reply by nomaam** [reply]
That does remove the error, however the actions and ledColor are not doing anything. I have also tried sending that data through withAndroidConfig through data and notifications without luck. 

Does this SDK support 'actions' and ledColor'?

**Reply by jeromegamez** [reply]
I couldn't find something related to actions in https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages?hl=en , but the LightSettings are available. The SDK doesn't support them yet, but I'll see if I can add support in the next days 🤞.

If you could provide with some official documentation for the actions, or how it is supported in other official SDKs I can give them a try as well 🤞

**Reply by nomaam** [reply]
That would be amazing! This SDK is great and helping a lot of people. 

Here is the Cordova implementation

    https://github.com/havesource/cordova-plugin-push/blob/master/docs/PAYLOAD.md#action-buttons

Here is some info on actions from android

    https://developer.android.com/training/notify-user/build-notification#Actions

"Direct reply" is another cool feature, where you can reply inside of the notification. Direct reply is listed just under actions in both the links mentioned above.

Please let me know if I can help you out with testing anything you come up with.

**Reply by jeromegamez** [reply]
I am not sure the actions can be added from a server-sent notification, at least I haven't found docs on them.

The Cordova Framework is a client side framework, so I believe actions have to be implemented on the client-side in general. I am fairly confident that unless those actions can be set via the FCM Rest API, you could still achieve this by sending a data message to the client (that is: a silent message without a notification payload) and act upon it on the client to trigger a visible notification from within your client app, including all the features that the device supports.

I imagine this being the preferred way in any case: LED Colors and actions are device-specific, if I'm not wrong? For example, there are Android Devices without LEDs and Android OS versions without action support, isn't it? (Not a rhetorical question, I honestly don't know 😅)

**Comment by jeromegamez** [comment]
I just checked the documentation again (I previously looked at it on my mobile phone) and am happy to be able to tell you that you can apply the LED/light settings now already. Now more explicitly referring to https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages?hl=en#androidnotification, you should be able to configure the light settings like this:

```php
$androidConfig = [
    'notification' => [
        'title' => 'Android specific notification title',
        'body' => 'Android specific notification body',
        // If set to true, use the Android framework's default LED light settings for the notification.
        // Default values are specified in config.xml. If default_light_settings is set to true and
        // light_settings is also set, the user-specified light_settings is used instead of the
        // default value.
        'default_light_settings' => false,
        // see https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#lightsettings
        'light_settings' => [
            // see https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#Color.SCHEMA_REPRESENTATION
            'color' => [
                'red' => 0,
                'green' => 0,
                'blue' => 255,
                'alpha' => 0
            ],
            'light_on_duration' => '3.5s',
            'light_off_duration' => '3.5s',

        ]
    ],
];

$message = CloudMessage::withTarget('token', $deviceToken)
    ->withAndroidConfig($androidConfig);

$messaging->send($message);
```

**Reply by nomaam** [reply]
Thank you for looking into this. I am unable to get this to work, perhaps my device doesn't support the feature. I am using a Samsung S9+, but I would think it would as there is a LED that lights up when I use other apps. I do not receive any type of error, just no lights come on. 

Have you had a chance to confirm this code as working?

**Comment by jeromegamez** [comment]
As for the actions - I just tested sending the actions example from the [Cordova docs you shared](https://github.com/havesource/cordova-plugin-push/blob/master/docs/PAYLOAD.md#action-buttons) after disabling the "one dimensional strings" check, and the Firebase API did not accept the message, as expected.

After researching the topic a little further, I stumbled upon https://stackoverflow.com/q/45006551/284325 which confirms that actions are not supported by the FCM Rest API and has a more detailed in-depth description of what I proposed earlier (JSON encoding on your backend, and decoding in the client) in its comments.

**Reply by nomaam** [reply]
Thanks for looking into this as well. Hopefully they will add support for this feature in the future. 

Take care.
